contracts/clients/src/BlsProvider.ts

```diff
@@ -0,0 +1,155 @@
+    this.verificationGatewayAddress = verificationGatewayAddress;
+  }
+
+  override async estimateGas(
+    transaction: Deferrable<ethers.providers.TransactionRequest>,
+  ): Promise<BigNumber> {
+    if (!transaction.to) {
+      throw new TypeError("Transaction.to should be defined.");
+    }
+
+    try {
+      const action: ActionData = {
+        ethValue: transaction.value?.toString() ?? "0",
+        contractAddress: transaction.to.toString(),
+        encodedFunction: transaction.data?.toString() ?? "0x",
+      };
+
+      const signer = this.getSigner();
+      const bundle = await signer.signBlsTransaction(action);
+
+      const gasEstimate = await this.aggregator.estimateFee(bundle);
+
+      return parseEther(gasEstimate.feeRequired);
+    } catch (error) {
+      throw new Error(`estimateGas() - an unexpected error occured: ${error}`);
+    }
+  }
+
+  override async sendTransaction(
+    signedTransaction: string | Promise<string>,
+  ): Promise<ethers.providers.TransactionResponse> {
+    throw new Error(
+      "sendTransaction() is not implemented. Call 'sendBlsTransaction()' instead.",
+    );
+  }
+
+  async sendBlsTransaction(
+    bundle: Bundle,
+    signer: BlsSigner,
+  ): Promise<ethers.providers.TransactionResponse> {
+    const result = await this.aggregator.add(bundle);
+
+    if ("failures" in result) {
+      throw new Error(JSON.stringify(result.failures));
+    }
+
+    // TODO: bls-wallet #375 Add multi-action transactions to BlsProvider & BlsSigner
+    // We're assuming the first operation and action constitute the correct values. We will need to refactor this when we add multi-action transactions
+    const actionData: ActionData = {
+      ethValue: bundle.operations[0].actions[0].ethValue.toString(),
+      contractAddress:
+        bundle.operations[0].actions[0].contractAddress.toString(),
+      encodedFunction:
+        bundle.operations[0].actions[0].encodedFunction.toString(),
+    };
+
+    return signer.constructTransactionResponse(
+      actionData,
+      result.hash,
+      signer.wallet.address,
+    );
+  }
+
+  override getSigner(addressOrIndex?: string | number): BlsSigner {
+    if (this.signer) {
+      return this.signer;
+    }
+
+    const signer = new BlsSigner(_constructorGuard, this, addressOrIndex);
+    this.signer = signer;
+    return signer;
+  }
+
+  override async getTransactionReceipt(
+    transactionHash: string | Promise<string>,
+  ): Promise<ethers.providers.TransactionReceipt> {
+    const resolvedTransactionHash = await transactionHash;
+    return this._getTransactionReceipt(resolvedTransactionHash, 1, 10);
+  }
+
+  override async waitForTransaction(
+    transactionHash: string,
+    confirmations?: number,
+    retries?: number,
+  ): Promise<ethers.providers.TransactionReceipt> {
+    return this._getTransactionReceipt(
+      transactionHash,
+      confirmations == null ? 1 : confirmations,
+      retries == null ? 10 : retries,
```

Comment:
Does `null` have special meaning here? Otherwise:
```suggestion
      confirmations ?? 1,
      retries ?? 10,
```